File: JavaCivitas/src/civitas/Tablero.java
package civitas;

import java.util.ArrayList;

/**
 *
 * @author Daniel
 */

public class Tablero {
    
    private int numCasillaCarcel;
    private ArrayList<Casilla> casillas;
    private int porSalida;
    private Boolean tieneJuez;
    
    // Constructor
    public Tablero(int indice_carcel){
        
        if (indice_carcel > 1 ) numCasillaCarcel = indice_carcel;
        else{
            
            numCasillaCarcel = 1;
            
        }

        casillas=new ArrayList<Casilla>();
        
        casillas.clear();

        casillas.add(new Casilla("Salida"));
        
        porSalida = 0;
        
        tieneJuez = false;         
    }
    
    private boolean correcto(){
        
        if(casillas.size() > numCasillaCarcel && tieneJuez){
            return true;
        }else{
            return false;
        }
    }
    
    private boolean correcto(int numCasillas){
        if(correcto() && casillas.size() > numCasillas && numCasillas >= 0) {
            return true;
        }else{
            return false;
        }
    }
    
    int getCarcel(){
        
        return numCasillaCarcel;
        
    }
    
    int getPorSalida(){
        
        if(porSalida > 0){
            int salir=porSalida;
            porSalida--;
            
            return salir;
        } 
        
        return porSalida;      
        
    }
    
    void añadeCasilla(Casilla casilla){
        Casilla Carcel = new Casilla("Carcel");
        if(casillas.size() == numCasillaCarcel){
            
            casillas.add(Carcel);
            
        }
        
        casillas.add(casilla);
        
        if (casillas.size() == numCasillaCarcel){
            
            casillas.add(Carcel);
            
        }       
        
    }
    
    void añadeJuez(){
        
        if(!tieneJuez) {
            casillas.add(new CasillaJuez(numCasillaCarcel,"Juez") );
            tieneJuez = true;
        }
        
    }
    
    Casilla getCasilla(int numCasilla){
        
        if(correcto(numCasilla)) {

            return casillas.get(numCasilla);


        }


        return casillas.get(numCasilla);

    }
    
    int nuevaPosicion(int actual, int tirada){
        


            int pos_final = actual + tirada;

            if (pos_final >= casillas.size()) {

                porSalida++;
                pos_final = (pos_final % casillas.size()) ;

            }
            return pos_final;

        

        
    }
    
    int calcularTirada (int origen, int destino){
        
        int tirada_nec = destino - origen;
        
        if(tirada_nec < 0){
            tirada_nec = tirada_nec+casillas.size();
        }
        
        return tirada_nec;
        
    }    
}

File: civitasruby/lib/tablero.rb
# encoding:utf-8
module Civitas
class Tablero
  require_relative ('casilla')
  attr_reader :numCasillaCarcel
  def initialize(indice)
    if(indice>=1)
      @numCasillaCarcel=indice
    else
      @numCasillaCarcel=1   
  end
  @casillas=Array.new
  @casillas<<Casilla.new("Salida")
 
  @porSalida=0
  @tieneJuez=false
  end


  def correcto(num_casilla=nil)
    
      if(num_casilla==nil)
       return @tieneJuez && @casillas.size > @numCasillaCarcel
      
      else
        if(@tieneJuez && @casillas.size > @numCasillaCarcel)
            return @casillas.size > num_casilla
          end   
      end
    end
 
  
  def  get_porsalida()
      aux=@porSalida
      if(@porSalida>0)
        @porSalida=@porSalida-1
        return aux
      else
        return @porSalida
      end  
  end
  
  def anade_casilla(casilla)
    if(@casillas.size==@numCasillaCarcel)
        @casillas<<Casilla.new("Cárcel")
        @casillas<<casilla
    else
      @casillas<<casilla
    end
  end
  
  def anade_juez()
    if(!@tieneJuez)
      @casillas<<CasillaJuez.new("Juez",@numCasillaCarcel)
      @tieneJuez=true
    end
  end
  
  
  def get_casilla(numCasilla)
    if(correcto(numCasilla))
      return @casillas[numCasilla]
    else
      return nil
    end
    
  end
  
  
  def nueva_posicion(actual,tirada)
    if(correcto())
      a=actual+tirada
      b=(actual+tirada)% @casillas.size
      if(a==b)
        return a
      else
        @poSalida=@porSalida+1
        return b
      end
    else
      return -1
    end
  end
  
  def calcular_tirada(origen,destino)
    a=destino-origen
    if(a<0)
      @porSalida=@porSalida+1
      return a+@casillas.size
    else
      return a
    end
    
  end
  
  
 
  
  private :correcto
  
  end
end